File: python/mirage/commands/spec/argmatch.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from dataclasses import dataclass
from typing import Literal

# One ARGMATCH candidate set. Each entry is one VALUE, spelled either as
# the single word that names it or as a tuple of alias words that all
# mean it. The grouping is not cosmetic: it is the whole of gnulib's
# ambiguity rule (two matching words inside one entry are one value and
# match; two matching entries are two values and are ambiguous), and it
# is what `argmatch_valid_block` renders one `  - ` line per, which is
# how GNU answers `sort --check=x` with `  - 'quiet', 'silent'` on one
# line and `  - 'diagnose-first'` on the next.
ArgmatchChoices = tuple[str | tuple[str, ...], ...]

# The two wordings a refused value picks between, spelled as the word the
# message carries (`ls: invalid argument 'zzz' for '--time'`).
ArgmatchKind = Literal["ambiguous", "invalid"]


@dataclass(frozen=True, slots=True)
class ArgmatchMatch:
    """A value that resolved to exactly one candidate value.

    Args:
        word (str): the canonical spelling of that value -- the first
            word of its class, which is the spelling the caller looks
            the meaning up under, so `--sort=non` answers `none` and
            `--time=use` answers `atime`.
    """
    word: str


@dataclass(frozen=True, slots=True)
class ArgmatchRefusal:
    """A value gnulib refuses, and which of its two wordings says so.

    Args:
        kind (ArgmatchKind): "ambiguous" when the value is a prefix of
            words spanning two or more values, "invalid" when it is a
            prefix of none.
    """
    kind: ArgmatchKind


ArgmatchResult = ArgmatchMatch | ArgmatchRefusal


def value_classes(choices: ArgmatchChoices) -> tuple[tuple[str, ...], ...]:
    """``choices`` as one tuple of alias words per value.

    A flat candidate tuple is a set of one-word values, which is what
    lets a spec's plain ``choices=`` reach the matcher with no change to
    its declaration.

    Args:
        choices (ArgmatchChoices): the candidates in declaration order.
    """
    grouped = ((choice, ) if isinstance(choice, str) else tuple(choice)
               for choice in choices)
    return tuple(group for group in grouped if group)


def argmatch(value: str, choices: ArgmatchChoices) -> ArgmatchResult:
    """gnulib's ``argmatch`` for one option value.

    The three answers of `argmatch (arg, arglist, vallist, valsize)`,
    in one pass, measured against coreutils 9.4:

    1. An EXACT word wins outright, even when it is a proper prefix of
       a longer candidate: `ls --quoting-style=shell` is `shell`, not an
       ambiguity with `shell-always`, and `ls --color=no` is `never`.
    2. Otherwise every word ``value`` is a prefix of matches. Plain byte
       prefix, case-sensitive (`ls --sort=NON` is invalid, neither
       ambiguous nor accepted), and a hyphen is an ordinary character
       (`--quoting-style=shell-a` resolves).
    3. Matching words are ACCEPTED when they all mean one value, and
       ambiguous only when they span two or more: `ls -l --time=a` hits
       `atime` and `access`, one value, exit 0, while `--time=c` hits
       `ctime` and `creation`, two values, exit 1.

    There is deliberately no empty-string branch, and adding one would
    be a bug. `""` is a prefix of every candidate, so it falls through
    rule 2 into rule 3 and is ambiguous exactly when the candidates span
    two or more values -- which is why `ls --sort=` and `wc --total=`
    answer `ambiguous argument ''` while a one-value set accepts it.
    gnulib does no such check either: its test is
    `strncmp (arglist[i], arg, 0) == 0`, true for every candidate.

    Args:
        value (str): the value as typed, never pre-escaped.
        choices (ArgmatchChoices): the candidates in declaration order,
            alias words of one value grouped into a tuple.

    Returns:
        ArgmatchResult: ``ArgmatchMatch`` carrying the canonical word,
            or ``ArgmatchRefusal`` carrying the wording GNU picks.
    """
    classes = value_classes(choices)
    for group in classes:
        if value in group:
            return ArgmatchMatch(group[0])
    matched = [
        group for group in classes if any(
            word.startswith(value) for word in group)
    ]
    if not matched:
        return ArgmatchRefusal("invalid")
    if len({group[0] for group in matched}) > 1:
        return ArgmatchRefusal("ambiguous")
    return ArgmatchMatch(matched[0][0])

File: typescript/packages/core/src/commands/spec/argmatch.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

/**
 * One ARGMATCH candidate set. Each entry is one VALUE, spelled either as
 * the single word that names it or as an array of alias words that all
 * mean it. The grouping is not cosmetic: it is the whole of gnulib's
 * ambiguity rule (two matching words inside one entry are one value and
 * match; two matching entries are two values and are ambiguous), and it
 * is what `argmatchValidBlock` renders one `  - ` line per, which is how
 * GNU answers `sort --check=x` with `  - 'quiet', 'silent'` on one line
 * and `  - 'diagnose-first'` on the next.
 *
 * `ArgmatchChoices` in argmatch.py is the twin.
 */
export type ArgmatchChoices = readonly (string | readonly string[])[]

/**
 * The two wordings a refused value picks between, spelled as the word the
 * message carries (`ls: invalid argument 'zzz' for '--time'`).
 */
export type ArgmatchKind = 'ambiguous' | 'invalid'

/**
 * What `argmatch` answers: the canonical word of the one value that
 * matched, or the wording GNU refuses the value with. Discriminated on
 * `matched` so a caller cannot read a refusal as a word, mirroring
 * Python's `ArgmatchMatch | ArgmatchRefusal`.
 */
export type ArgmatchResult =
  | { readonly matched: true; readonly word: string }
  | { readonly matched: false; readonly kind: ArgmatchKind }

/**
 * `choices` as one array of alias words per value.
 *
 * A flat candidate list is a set of one-word values, which is what lets a
 * spec's plain `choices` reach the matcher with no change to its
 * declaration.
 */
export function valueClasses(choices: ArgmatchChoices): readonly (readonly string[])[] {
  return choices
    .map((choice) => (typeof choice === 'string' ? [choice] : choice))
    .filter((group) => group.length > 0)
}

/**
 * gnulib's `argmatch` for one option value.
 *
 * The three answers of `argmatch (arg, arglist, vallist, valsize)`, in
 * one pass, measured against coreutils 9.4:
 *
 * 1. An EXACT word wins outright, even when it is a proper prefix of a
 *    longer candidate: `ls --quoting-style=shell` is `shell`, not an
 *    ambiguity with `shell-always`, and `ls --color=no` is `never`.
 * 2. Otherwise every word `value` is a prefix of matches. Plain byte
 *    prefix, case-sensitive (`ls --sort=NON` is invalid, neither
 *    ambiguous nor accepted), and a hyphen is an ordinary character
 *    (`--quoting-style=shell-a` resolves).
 * 3. Matching words are ACCEPTED when they all mean one value, and
 *    ambiguous only when they span two or more: `ls -l --time=a` hits
 *    `atime` and `access`, one value, exit 0, while `--time=c` hits
 *    `ctime` and `creation`, two values, exit 1.
 *
 * There is deliberately no empty-string branch, and adding one would be
 * a bug. `''` is a prefix of every candidate, so it falls through rule 2
 * into rule 3 and is ambiguous exactly when the candidates span two or
 * more values -- which is why `ls --sort=` and `wc --total=` answer
 * `ambiguous argument ''` while a one-value set accepts it. gnulib does
 * no such check either: its test is `strncmp (arglist[i], arg, 0) == 0`,
 * true for every candidate.
 *
 * `argmatch` in argmatch.py is the twin.
 */
export function argmatch(value: string, choices: ArgmatchChoices): ArgmatchResult {
  const classes = valueClasses(choices)
  for (const group of classes) {
    if (group.includes(value)) return { matched: true, word: group[0] ?? '' }
  }
  const matched = classes.filter((group) => group.some((word) => word.startsWith(value)))
  const first = matched[0]
  if (first === undefined) return { matched: false, kind: 'invalid' }
  if (new Set(matched.map((group) => group[0])).size > 1) {
    return { matched: false, kind: 'ambiguous' }
  }
  return { matched: true, word: first[0] ?? '' }
}
